source/common/upstream/eds.cc

```diff
@@ -143,10 +178,75 @@ void EdsClusterImpl::onConfigUpdate(const std::vector<Config::DecodedResourceRef
     assignment_timeout_->enableTimer(std::chrono::milliseconds(stale_after_ms));
   }
 
-  BatchUpdateHelper helper(*this, cluster_load_assignment);
+  // TODO(adisuissa) : Add a test to verify that a cluster can switch from warming -> active ->
+  // warming by adding a new LEDS locality, and that the Envoy cluster isn't updated until we
+  // receive the new LEDS data.
+
+  // Compare the current set of LEDS localities (localities using LEDS) to the one received in the
+  // update. A LEDS locality can either be added, removed, or kept. If it is added we add a
+  // subscription to it, and if it is removed we delete the subscription.
+  LedsConfigSet cla_leds_configs;
+
+  for (const auto& locality : cluster_load_assignment.endpoints()) {
+    if (locality.has_leds_cluster_locality_config()) {
+      cla_leds_configs.emplace(locality.leds_cluster_locality_config());
+    }
+  }
+
+  // Remove the LEDS localities that are not needed anymore.
+  absl::erase_if(leds_localities_, [&cla_leds_configs](const auto& item) {
+    auto const& [leds_config, _] = item;
+    // Returns true if the leds_config isn't in the cla_leds_configs
+    return cla_leds_configs.find(leds_config) == cla_leds_configs.end();
+  });
+
+  // Add all the LEDS localities that are new.
+  for (const auto& leds_config : cla_leds_configs) {
+    if (leds_localities_.find(leds_config) == leds_localities_.end()) {
+      ENVOY_LOG(debug, "Found new LEDS config in EDS onConfigUpdate() for cluster {}: {}",
+                cluster_name_, leds_config.DebugString());
+
+      // Create a new LEDS subscription and add it to the subscriptions map.
+      LedsSubscriptionPtr leds_locality_subscription = std::make_unique<LedsSubscription>(
+          leds_config, cluster_name_, factory_context_, /*nullptr,*/ [&]() {
+            // Called upon an update to the locality
+            // TODO(adisuissa): here we know the exact locality that is now active, so we can
+            // optimize this process by moving the locality from warming set ot active set, and call
+            // the update if the warming set is empty.
+            if (validateAllLedsActive()) {
```

Comment:
Does the design here support transitive warming flows?

E.g. when a new cluster is added, it will warm on its EDS update. We need to delay the warming of the EDS update until all the dependent LEDS updates have occurred.